controller/node_controller.go

```diff
@@ -731,3 +765,169 @@ func (nc *NodeController) syncNodeStatus(pod *v1.Pod, node *longhorn.Node) error
 
 	return nil
 }
+
+func (nc *NodeController) syncEnginePods(node *longhorn.Node) error {
+	engineImages, err := nc.ds.ListEngineImages()
+	if err != nil {
+		return err
+	}
+	for id, image := range engineImages {
+		enginePod, err := nc.ds.GetNodeEnginePod(nc.controllerID, image.Name, types.EngineManagerType)
+		if err != nil {
+			return errors.Wrapf(err, "cannot get engine manager pod for engine image %v", id)
+		}
+		if enginePod == nil {
+			enginePodSpec := nc.createNodeEngineManagerPodSpec(image)
+			if err = nc.ds.CreateNodeEnginePod(enginePodSpec); err != nil {
+				return errors.Wrapf(err, "failed to create engine manager pod for engine image %v", id)
+			}
+			logrus.Infof("Created engine manager pod on node %v for engine image %v (%v)", nc.controllerID, id, image.Spec.Image)
+		}
+
+		replicaPod, err := nc.ds.GetNodeEnginePod(nc.controllerID, image.Name, types.ReplicaManagerType)
+		if err != nil {
+			return errors.Wrapf(err, "cannot get replica launcher pod for engine image %v", id)
+		}
+		if replicaPod == nil && len(node.Spec.Disks) > 0 {
```

Comment:
We're intending on deploying the `Replica Manager Pod` only if the `Node` has `Disks`, since otherwise there's nowhere to store the `Replicas`.

However, what happens if the `Disks` get removed after this `Pod` has already been created? How should we address this?

Reply:
Yes, we need to address this. The replica pod lifecycle is bound with the availability of the disk. One disk can only be removed if there is no replica stored in it, so it should be safe to remove the replica pod as well.

Reply:
Adding logic to delete the `Replica Pod` once no `Disks` remain on the `Node`.